bench/irmin-pack/tree.ml

```diff
@@ -0,0 +1,351 @@
+            Store.Tree.list t.tree [] >>= fun _ ->
+            let info = Irmin.Info.v ~date ~author:"Tezos" message in
+            let parents =
+              match t.prev_commit with None -> [] | Some p -> [ p ]
+            in
+            Store.Commit.v repo ~info ~parents t.tree >|= fun c ->
+            t.prev_commit <- Some (Store.Commit.hash c);
+            Store.Tree.clear t.tree)
+      operations
+
+  let add_commits repo () =
+    let t = { tree = Store.Tree.empty; prev_commit = None } in
+    let rec array_iter_lwt i =
+      if i >= Array.length !commits then Lwt.return_unit
+      else
+        let operations = !commits.(i) in
+        add_operations t repo operations i >>= fun () -> array_iter_lwt (i + 1)
+    in
+    array_iter_lwt 0
+end
+
+let () = Random.self_init ()
+
+type config = {
+  ncommits : int;
+  mutable ncommits_log : int;
+  depth : int;
+  nchain_trees : int;
+  width : int;
+  nlarge_trees : int;
+  operations_file : string;
+  root : string;
+  mode : [ `Tezos_log | `Chains | `Large | `All ];
+}
+
+module Benchmark = struct
+  type result = { time : float; size : int }
+
+  let run config f =
+    with_timer f >|= fun (time, _) ->
+    let size = FSHelper.get_size config.root in
+    { time; size }
+
+  let pp_results fmt result =
+    Format.fprintf fmt "Total time: %f@\nSize on disk: %d M" result.time
+      result.size
+end
+
+module Hash = Irmin.Hash.SHA1
+
+module Bench_suite (Conf : sig
+  val entries : int
+  val stable_hash : int
+end) =
+struct
+  module Store =
+    Irmin_pack.Make (Conf) (Irmin.Metadata.None) (Irmin.Contents.String)
+      (Irmin.Path.String_list)
+      (Irmin.Branch.String)
+      (Hash)
+
+  let init_commit repo =
+    Store.Commit.v repo ~info:(info ()) ~parents:[] Store.Tree.empty
+
+  module Trees = Generate_trees (Store)
+  module Trees_log = Generate_trees_from_log (Store)
+
+  let checkout_and_commit repo prev_commit f =
+    Store.Commit.of_hash repo prev_commit >>= function
+    | None -> Lwt.fail_with "commit not found"
+    | Some commit ->
+        let tree = Store.Commit.tree commit in
+        f tree >>= fun tree ->
+        Store.Commit.v repo ~info:(info ()) ~parents:[ prev_commit ] tree
+
+  let add_commits repo ncommits f () =
+    init_commit repo >>= fun c ->
+    let rec aux c i =
+      if i >= ncommits then Lwt.return c
+      else
+        checkout_and_commit repo (Store.Commit.hash c) f >>= fun c' ->
+        aux c' (i + 1)
+    in
+    aux c 0 >|= ignore
+
+  let run mode config =
+    let conf = Irmin_pack.config ~readonly:false ~fresh:true config.root in
+    Store.Repo.v conf >>= fun repo ->
+    (match mode with
+    | `Large ->
+        Trees.add_large_trees config.width config.nlarge_trees
+        |> add_commits repo config.ncommits
+    | `Chains ->
+        Trees.add_chain_trees config.depth config.nchain_trees
+        |> add_commits repo config.ncommits
+    | `Tezos_log -> Trees_log.add_commits repo
```

Comment:
That's great. The `tezos-log` mode feels like a real world benchmark.

I'm able to simulate the 13315 commits in 70 seconds. How long does it take for tezos to generate that many commits?